orttraining/orttraining/training_ops/cpu/communication/recv.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/communication/recv.h"
+
+#include <mpi.h>
+
+#include "orttraining/training_ops/communication_common.h"
+#include "orttraining/core/framework/mpi_context.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+void Recv::ReceiveData(
+    const int num_tensors,
+    std::vector<Tensor*> received_tensors,
+    const int src,
+    const size_t aggregated_aligned_tensor_bytes,
+    std::vector<char>& buffer) const {
+  buffer.reserve(aggregated_aligned_tensor_bytes);
+  CommInfo_t info_data{buffer.data(),
+                       static_cast<int>(aggregated_aligned_tensor_bytes),
+                       src,
+                       static_cast<int>(tag_)};
+
+  MPI_CHECK(MPI_Recv(
+      info_data.buffer, info_data.size, MPI_CHAR,
+      info_data.rank, info_data.tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE));
+
+  // Copy tensors from buffer to outputs.
+  size_t tensor_offset_in_bytes = 0;
+  for (int i = 0; i < num_tensors; ++i) {
+    Tensor* tensor = received_tensors[i];
+
+    // Find the next aligned offset in the tensor buffer to meet alignment requirement
+    tensor_offset_in_bytes = GetAggregatedAlignedAddress(tensor_offset_in_bytes);
+
+    // Copy data out from buffer.
+    memcpy(tensor->MutableDataRaw(), buffer.data() + tensor_offset_in_bytes, tensor->SizeInBytes());
+
+    tensor_offset_in_bytes += tensor->SizeInBytes();
+  }
+}
+
+ONNX_OPERATOR_KERNEL_EX(
+    Recv,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeDefault>(0)  /* CPU variable */
+        .InputMemoryType<OrtMemTypeDefault>(1)  /* CPU variable */
+        .OutputMemoryType<OrtMemTypeDefault>(0) /* CPU variable */
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+    Recv);
+
+Status Recv::Compute(OpKernelContext* ctx) const {
```

Comment:
There is a lot in common here with the CUDA/GPU implementation.  How far can we avoid duplication by extracting some common helper functions into communication_common.h?  This separation will help in the future if we need to distinguish NVIDIA and AMD implementations of the communication primitives (if only at the level of how we manage profiling).

Reply:
There are some common code for both CPU and CUDA/GPU and the difference mainly at profiling and how we allocate the memory. Besides the profiling macro is interleaving the CUDA/GPU code, the main reason I prefer to keep them separate is that, while the CPU implementation can stay static as a reference, the CUDA/GPU implementation could change in the future -- we might consider remove the tensor aggregation for performance optimization and to support skip stage send/recv. And that is the main chunk of the function.

I am not sure how high the priority is for this change. If you feel like this is not in our near plan, I can merge them first for now. (re-using code fan)